csrc/alias_analysis.cpp

```diff
@@ -429,15 +432,21 @@ bool okToRelayout(
 } // namespace
 
 void AliasAnalysisResult::finalize(
-    const bool can_override_empty_allocation_domain) {
+    const bool can_override_empty_allocation_domain,
+    const bool allow_output_alias_intermediate) {
   for (auto [alias, root_and_layout] : alias_to_source_) {
     auto [root, preferred_layout] = root_and_layout;
-    // Walks up the `alias_to_source_` chain.
-    while (root != nullptr && !root->isFusionInput() &&
-           !root->isFusionOutput()) {
-      const auto i = alias_to_source_.find(root);
-      root = (i == alias_to_source_.end() ? nullptr : i->second.first);
+    // If allow output to alias intermediate, then we don't need to walk up
+    // the chain to find an input or output root. It changes the last reshape
```

Comment:
I'd hope it's backed by numbers, but a more reasonable approach seems to be walk up until we reach a tensor that are known to be global, e.g., when outputInterferingReduction returns true or, not in this PR, the input of slice or pad.

Reply:
A little confusion about `a tensor that are known to be global`. Here I am trying to let the output tensor alias to a tensor that is not in global memory (not input or output, but intermediate), then further code will segment this `output tensor = someOp(an intermediate tensor)` to a no-op.

Here `outputInterferingReduction` depends on the whole `fusion` not a single tensor. So I didn't move it into the walk up process. But it looks like we should also check alias in `alias_to_source_` is actually defined by a view op. Then we don't need to walk up. So the code is changed as:
(1) ` stop_at_view =  may_alias_intermediate && outputInterferingReduction(fusion)` is moved to `finalize()`
(2) only walk up if `if (!isOpsToStop(alias->definition(), stop_at_view))`